sentry-android-core/src/main/java/io/sentry/android/core/NetworkBreadcrumbsIntegration.java

```diff
@@ -0,0 +1,230 @@
+      final boolean registered =
+          ConnectivityChecker.registerNetworkCallback(
+              context, logger, buildInfoProvider, networkCallback);
+
+      // The specific error is logged in the ConnectivityChecker method
+      if (!registered) {
+        networkCallback = null;
+        logger.log(SentryLevel.DEBUG, "NetworkBreadcrumbsIntegration not installed.");
+        return;
+      }
+      logger.log(SentryLevel.DEBUG, "NetworkBreadcrumbsIntegration installed.");
+      addIntegrationToSdkVersion();
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (networkCallback != null) {
+      ConnectivityChecker.unregisterNetworkCallback(
+          context, logger, buildInfoProvider, networkCallback);
+      logger.log(SentryLevel.DEBUG, "NetworkBreadcrumbsIntegration remove.");
+    }
+    networkCallback = null;
+  }
+
+  @SuppressLint("ObsoleteSdkInt")
+  @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+  static final class NetworkBreadcrumbsNetworkCallback extends ConnectivityManager.NetworkCallback {
+    final @NotNull IHub hub;
+    final @NotNull BuildInfoProvider buildInfoProvider;
+
+    @Nullable Network currentNetwork = null;
+
+    @Nullable NetworkCapabilities lastCapabilities = null;
+
+    NetworkBreadcrumbsNetworkCallback(
+        final @NotNull IHub hub, final @NotNull BuildInfoProvider buildInfoProvider) {
+      this.hub = Objects.requireNonNull(hub, "Hub is required");
+      this.buildInfoProvider =
+          Objects.requireNonNull(buildInfoProvider, "BuildInfoProvider is required");
+    }
+
+    @Override
+    public void onAvailable(final @NonNull Network network) {
+      if (network.equals(currentNetwork)) {
+        return;
+      }
+      final Breadcrumb breadcrumb = createBreadcrumb("networkAvailable");
+      hub.addBreadcrumb(breadcrumb);
+      currentNetwork = network;
+      lastCapabilities = null;
+    }
+
+    @Override
+    public void onCapabilitiesChanged(
+        final @NonNull Network network, final @NonNull NetworkCapabilities networkCapabilities) {
+      if (!network.equals(currentNetwork)) {
+        return;
+      }
+      final @Nullable NetworkBreadcrumbConnectionDetail connectionDetail =
+          getNewConnectionDetails(lastCapabilities, networkCapabilities);
+      if (connectionDetail == null) {
+        return;
+      }
+      lastCapabilities = networkCapabilities;
+      final Breadcrumb breadcrumb = createBreadcrumb("networkCapabilitiesChanged");
+      breadcrumb.setData("download_bandwidth", connectionDetail.downBandwidth);
+      breadcrumb.setData("upload_bandwidth", connectionDetail.upBandwidth);
+      breadcrumb.setData("vpn_active", connectionDetail.isVpn);
+      breadcrumb.setData("network_type", connectionDetail.type);
+      if (connectionDetail.signalStrength != 0) {
+        breadcrumb.setData("signal_strength", connectionDetail.signalStrength);
+      }
+      Hint hint = new Hint();
+      hint.set("data", connectionDetail);
```

Comment:
```suggestion
      hint.set(ANDROID_NETWORK_CAPABILITIES, connectionDetail);
```

I believe we should create a key in `TypeCheckHint` to align with other hints, something like 
```
public static final String ANDROID_NETWORK_CAPABILITIES = "android:networkCapabilities";
```

Reply:
this can be used by our users in BeforeBreadcrumbCallback to extract this data and drop/adjust the recorded breadcrumb

Reply:
yes, that should be ok